Move the notch computation from the FAB to the BAB.

amirh opened the pull request:
The notch in the BAB (bottom action bar) for the FAB (floating action button) was previously kept as part of the FAB's implementation. This was done to keep the shape of the FAB and the shape of the notch coupled.
That approach resulted in a somewhat complex and 'non Fluttery' mechanism for propagating the notch computation from the FAB to the BAB.

This CL uncouples the FAB and the notch computation.
With the new API the BAB computes its overall shape including the notch using a NotchedShape delegate.

This includes multiple breaking changes:
  * Scaffold.setFloatingActionButtonNotchFor is deleted.
  * The ComputeNotch type is deleted.
  * The hasNotch property of BottomAppBar is deleted.
  * The notchMargin property of FloatingActionButton is deleted.

Quick migration guide from the previous API:

| Previous API | New API |
| ------------------|-------------|
| BottomAppBar(hasNotch: false) | BottomAppBar() |
| Using a FloatingActionButton with: BottomAppBar() / BottomAppBar(hasNotch: true) | BottomAppBar(shape: CircularNotchedRectangle()) |
| Scaffold.setFloatingActionButtonNotchFor(..) | No longer supported |

amirh committed:
Move the notch computation from the FAB to the BAB.

The notch in the bottom action bar for the floating action button was previously
kept as part of the FAB's implementation to keep the shape of the FAB and the
shape of the notch coupled.
That approach resulted in a somewhat complex and 'non Fluttery'
mechanism for propagating the notch computation from the FAB to the BAB.

This CL uncouples the FAB and the notch computation, by making the notch
computer an independent property of the BottomAppBar.

This includes multiple breaking changes:
  * Scaffold.setFloatingActionButtonNotchFor is deleted.
  * The ComputeNotch type is deleted.
  * The hasNotch property of BottomAppBar is deleted.
  * The notchMargin property of FloatingActionButton is deleted.

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+/// Computes notches in the outline of a shape.
+///
+/// Typically used to compute notches by a [BottomAppBar].
+abstract class NotchComputer {
```
Although I appreciate the symmetry ComputeNotch and NotchComputer, calling a class FooComputer seems a little odd.  The "computer" suffix sounds like a noun.  Like the Kitchen Computer (https://en.wikipedia.org/wiki/Honeywell_316#Kitchen_Computer) :-).

Why not just call the class Notch?

amirh replied:
done

amirh replied:
Also renamed notch_computers.dart to notches.dart (and the test file accordingly)

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+  /// not end with a [Path.close]. The returned [Path] is built under the
+  /// assumption it will be added to an existing path that is at the `start`
+  /// coordinates using [Path.addPath].
+  Path compute(Rect host, Rect guest, Offset start, Offset end);
```
It looks like most of the existing Flutter methods that return a path are called getFoo: CustomClipper.getClip(), ShapeBorder.getOuterPath, ShapeBorder.getInnerPath

Maybe call this getPath()?

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+/// Computes smooth circular notches.
+///
+/// The notches computed by this class are paths that smoothly connect a top
+/// horizontal edge with a circle circumference.
```
circle circumference => circular arc

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+/// horizontal edge with a circle circumference.
+///
+/// See also: [compute].
+class CircularNotchComputer implements NotchComputer {
```
I'd prefer CircularNotch implements Notch

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+  /// shape is circular.
+  final double notchMargin;
+
+  /// Returns a path for a smooth circular notch in a horizontal top edge of a shape.
```
I think you could be more imperative here: Returns a path that defines a circular notch in the top edge of the [host] shape.

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+  /// that will be connected by the returned path.
+  @override
+  Path compute(Rect host, Rect guest, Offset start, Offset end) {
+    // The guest's shape is a circle bounded by the guest rectangle.
```
Perhaps you should add some sanity check assertions here, host and guest rectangles are not empty.

amirh replied:
Is it possible that a empty host/guest is actually a valid scenario? (e.g maybe during some scale transition?) are you sure we want to crash in this case (even if only in debug mode)?

HansMuller replied:
Good point. I agree that it's better to just gracefully handle the exceptional cases (empty guest or host).

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+      ..lineTo(end.dx, end.dy);
+  }
+
+  bool _notchAssertions(Rect host, Rect guest, Offset start, Offset end,
```
The messages generated here should be in terms of host and guest, not FAB

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+
+/// Computes smooth circular notches.
+///
+/// The notches computed by this class are paths that smoothly connect a top
```
I think it's important to explain what the notch's radius is: guest.longestSide + notchMargin

amirh replied:
done.
I didn't mention longestSide as the guest shape is assumed to be a circle (hence the guest bounding box is assumed to be a square).

I added a comment and assertion for this.

HansMuller replied:
Sounds good

HansMuller commented on packages/flutter/lib/src/material/notch_computers.dart:
```diff
@@ -0,0 +1,158 @@
+    this.notchMargin = 4.0,
+  });
+
+  /// Minimal margin that will be kept from the guest. Assuming that the guest's
```
It might be clearer to just say that we're expanding the guest's implicit radius by notchMargin. You'd need to explain how we come up with the guest's implicit radius of course.

HansMuller reviewed (commented):
Just starting to review this. 

I think the overall direction is good and it's probably worth warning developers about the incompatible changes now.

amirh committed:
review comments

amirh commented:
PTAL

HansMuller commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,174 @@
+  /// const constructors so that they can be used in const expressions.
+  const Notch();
+
+  /// Returns a path for a notch in the outline of a shape.
```
Returns a [Path] that defines a cutout or "notch" in the host shape that accommodates the guest shape.

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,174 @@
+
+  /// Returns a path for a notch in the outline of a shape.
+  ///
+  /// The path makes a notch in the host shape that can contain the guest shape.
```
If you update the previous line, you could leave this one out.

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,174 @@
+    this.notchMargin = 4.0,
+  });
+
+  /// Minimal margin that will be kept from the guest. Assuming that the guest's
```
The margin between the circular guest and the host's circular notch.

amirh replied:
You suggestion is easier to parse.
Though if we want to be accurate we need to say it's the minimal margin (as around the rounded corners the margin might be bigger).

I'm fine with both, let me know which one you prefer.

HansMuller replied:
I think is OK to be a little less precise here. You could say that we're computing a [Minkowski sum](http://hansmuller-webkit.blogspot.com/2013/04/growing-and-shrinking-polygons-round-one.html) but that wouldn't really help anyone :-)

amirh replied:
Wow :smiley:

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,174 @@
+  /// The `start` and `end` arguments are points on the outline of the host shape
+  /// that will be connected by the returned path.
+  @override
+  Path getPath(Rect host, Rect guest, Offset start, Offset end) {
```
Tests should verify getPath(h, g, start, end) produces the same path as getPath(h, g, end, start)

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,174 @@
+        'The notch\'s path start point: $start is not in the top edge of $host'
+      );
+
+    if (guest.center.dx - notchRadius < start.dx)
```
Why not just have CircularNotch swap start and end if necessary?

amirh replied:
done

HansMuller commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,174 @@
+
+  bool _notchAssertions(Rect host, Rect guest, Offset start, Offset end,
+    double fabRadius, double notchRadius, double notchMargin) {
+    if (end.dy != host.top)
```
Don't you need to verify that the start/end X coordinates are within the host rectangle?  And if so, the API doc should say as much

amirh replied:
The doc says:
```
  /// The `start` and `end` arguments are points on the outline of the host shape
  /// that will be connected by the returned path.
```

Added two more assertion that left and right are not to the left or right of the host.

HansMuller reviewed (approved):
LGTM

An additional PR that added the diamond notch and a stadium notch (for extended FABs) would be good

amirh committed:
review comments followup

Hixie commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,187 @@
+  /// and may contain multiple subpaths. The returned path ends at `end` and does
+  /// not end with a [Path.close]. The returned [Path] is built under the
+  /// assumption it will be added to an existing path that is at the `start`
+  /// coordinates using [Path.addPath].
```
these docs switch between describing the contract as something that has to be implemented ("may") and describing the contract as something that can be consumed ("is"). We should pick one and stick to it (probably the former style, since people will be implementing this not using it, typically).

Hixie commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  bool _notchAssertions(Rect host, Rect guest, Offset left, Offset right,
+    double fabRadius, double notchRadius, double notchMargin) {
```
should this be in the superclass? or is it specific to this subclass?

We should document these requirements as things that implementations of getPath can assume.

Hixie commented on packages/flutter/lib/src/material/notches.dart:
```diff
@@ -0,0 +1,187 @@
+    if (right.dy != host.top)
+      throw new FlutterError(
+        'The notch of the guest must start and end at the top edge of the host.\n'
+        'The notch\'s path right point: $right is not in the top edge of $host'
```
missing trailing period

Hixie commented:
Where does a developer configure the shape of the FAB and the notch in the BAB? Are they roughly in the same place?

amirh commented:
Regarding configuring the shape of the FAB and the BAB's notch, they are no longer in the same place: FAB's shape is controlled by the FAB, and the notch shape is configured with a BAB parameter.
(if we submit this PR as is we basically decided to make the code simpler and pay by keeping the shapes configuration in different places).

I guess we could make both the BAB and FAB default to a shape from the Theme, and then at least in the default case the shapes will be configured roughly in the same place.

Happy to hear thoughts on this tradeoff, or if you have better suggestions.

CC @HansMuller.

Hixie commented:
If we can find a way to bring those two places together I think that'd be ideal from a developer perspective.

amirh commented:
That's what we currently do (before this CL): both shapes are configured from the FAB.
If we want to keep that approach we can probably still keep the new Notch and CircularNotch classes and pass them as parameters to the FAB.

The downside will be that we keep the `Scaffold.setFloatingActionButtonNotchFor` and the complexity it adds to `Scaffold`.

Hixie commented:
Oh I don't mean that the BAB would read from the FAB, I just mean like something like:

```dart
new Theme(
  fabShape: Circle,
  babNotch: CircleNotch,
),
```

(obviously not those actual names)

amirh commented:
Ohh I think that's what I meant when I said we could the make the FAB and the BAB notch shape default to Theme values.
But developers will still be able to directly set the shapes in unrelated places (unless you think to make it only configurable through the theme?)

Hixie commented:
yeah i think in general anything doable by theme should be available as an argument too.

amirh commented:
Following f2f discussion with @Hixie we're going to try and redo the entire notch API leveraging Path.combine.

Stay tuned for a different PR...
(I'm keeping this PR in case we missed something and the other approach fails)

amirh committed:
Compute both the entire BAB shape including the botch by a NotchedShape.

amirh commented:
Following more API design discussions I changed the BAB to take in a NotchedShape object which computes complete outline of the BAB (including the notch).

I'm pretty happy with the result, it simplified the code even more, and simplified the tests (the tests that are now in `notched_shapes_test.dart` were more complex before).

TBD in a following PR: add a `Path.combine` based `NotchedShape` implementation..

Hixie commented on packages/flutter/lib/src/painting/notched_shapes.dart:
```diff
@@ -0,0 +1,106 @@
+/// Typically used as the outline of a 'host' widget to make a notch that
+/// accomodates a 'guest' widget. e.g the [BottomAppBar] may have a notch to
+/// accomodate the [FloatingActionBar].
+abstract class NotchedShape {
```
maybe have a see-also link to Shape in case anyone got lost when reading the docs

amirh replied:
done

amirh replied:
(I assume you meant ShapeBorder)

Hixie commented on packages/flutter/lib/src/painting/notched_shapes.dart:
```diff
@@ -0,0 +1,106 @@
+}
+
+/// A rectangle with a smooth circular notch.
+class CircularNotchedRectangle implements NotchedShape {
```
consider calling this something like "DimpleNotchedRectangle" or some such, to make it clearer what the difference is between this and a semicircular notched rectangle.

amirh replied:
Naming here is pretty hard...
For a few moments I thought that once we expose the curve computation parameters (s1 and s2) we will avoid this naming problem, as one could configure this to be equivalent to the single arc notch, but there's actually no such setting for s1 and s2.

Another option would be to add a `smooth` boolean parameter to this class, but that's also not ideal as when it's false the s1/s2 parameters (which I plan to expose anyway) will be meaningless.

@HansMuller maybe you have some thoughts here?

HansMuller replied:
Maybe generate a simple circular notch if the s1/s2 parameters are null but have them default to values that produce a dimple.

amirh replied:
I'm ok with that - if either s1 or s2 are null we'll just default to the single arc notch. An this thing stays the CircularNotchedRectangle.

HansMuller replied:
Shouldn't there be s1/s2 constructor parameters?

Hixie commented on packages/flutter/lib/src/painting/notched_shapes.dart:
```diff
@@ -0,0 +1,106 @@
+    // Segment C - a Bezier curver from segment B back to the host's top edge.
+    //
+    // A detailed explanation and the derivation of the formulas below is
+    // available at: https://goo.gl/Ufzrqn
```
that document is awesome

amirh replied:
Thanks :smile:

Hixie commented on packages/flutter/test/material/bottom_app_bar_test.dart:
```diff
@@ -86,15 +90,15 @@ void main() {
 
-  // This is a regression test for a bug we had where toggling hasNotch
+  // This is a regression test for a bug we had where toggling the notch on/off
   // will crash, as the shouldReclip method of ShapeBorderClipper or
```
will->would

amirh replied:
done

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -71,16 +72,17 @@ class BottomAppBar extends StatefulWidget {
 
-  /// Whether to make a notch in the bottom app bar's shape for the floating
-  /// action button.
+  /// The notch that is used to make a notch for the floating action
```
this sentence doesn't really make sense any more. The first noun is probably wrong? Maybe "The delegate"?

amirh replied:
Delegate would make sense, but [hopefully] I simplified the entire sentence a bit.

Hixie commented on packages/flutter/lib/src/material/bottom_app_bar.dart:
```diff
@@ -71,16 +72,17 @@ class BottomAppBar extends StatefulWidget {
-  /// [ScaffoldGeometry.floatingActionButtonNotch] to make a notch along its
-  /// top edge, where it is overlapped by the
-  /// [ScaffoldGeometry.floatingActionButtonArea].
+  /// If null there will be no notch in the bottom app bar.
```
...and the app bar will be rectangular?

amirh replied:
done

amirh committed:
review comments followup

HansMuller commented on examples/flutter_gallery/lib/demo/material/bottom_app_bar_demo.dart:
```diff
@@ -470,14 +465,18 @@ class _DiamondFabState extends State<_DiamondFab> {
     final double notchToCenter =
-      intersection.height * (marginedGuest.height / 2.0)
-      / (marginedGuest.width / 2.0);
+      intersection.height * (guest.height / 2.0)
```
Even if it's impossible for guest.height to be 0 here, might be best to assert as much before dividing by guest.height.

amirh replied:
Added an assertion that guest.width is greater than 0.

HansMuller reviewed (approved):
LGTM

amirh commented:
I prefer leaving s1/s2 to a separate PR to keep this one more focused.

amirh committed:
review comments followup

amirh committed:
remove unnecessary dependency on rendering
